libcrux-ml-dsa/src/simd/portable/encoding/commitment.rs

```diff
@@ -1,40 +1,158 @@
+    let coefficient1 = coefficients[1] as u8;
+    (coefficient1 << 4) | coefficient0
+}
 
-                    serialized[i] = (coefficient1 << 4) | coefficient0;
-                }
-            }
-            ()
-        }
+#[inline(always)]
+#[hax_lib::requires(fstar!(r"Seq.length ${coefficients} == 4 /\ (forall i. bounded (Seq.index ${coefficients} i) 6)"))]
+#[hax_lib::ensures(|out| fstar!(r"
+let inp = bit_vec_of_int_t_array #I32 #(mk_usize 4) ${coefficients} 6 in
+let out = bit_vec_of_int_t_array #U8 (MkSeq.create3 ${out}) 8 in
+forall (i: nat {i < 8 * 3}). inp i == out i
+"))]
+fn encode_6(coefficients: &[i32]) -> (u8, u8, u8) {
+    let coefficient0 = coefficients[0] as u8;
+    let coefficient1 = coefficients[1] as u8;
+    let coefficient2 = coefficients[2] as u8;
+    let coefficient3 = coefficients[3] as u8;
 
-        6 => {
-            // The commitment has coefficients in [0,43] => each coefficient occupies
-            // 6 bits.
-            cloop! {
-                for (i, coefficients) in simd_unit.values.chunks_exact(4).enumerate() {
-                    let coefficient0 = coefficients[0] as u8;
-                    let coefficient1 = coefficients[1] as u8;
-                    let coefficient2 = coefficients[2] as u8;
-                    let coefficient3 = coefficients[3] as u8;
+    (
+        (coefficient1 << 6) | coefficient0,
+        (coefficient2 << 4) | coefficient1 >> 2,
+        (coefficient3 << 2) | coefficient2 >> 4,
+    )
+}
 
-                    serialized[3 * i] = (coefficient1 << 6) | coefficient0;
-                    serialized[3 * i + 1] = (coefficient2 << 4) | coefficient1 >> 2;
-                    serialized[3 * i + 2] = (coefficient3 << 2) | coefficient2 >> 4;
-                }
-            }
-            ()
+#[inline(always)]
+#[hax_lib::requires(fstar!(r"Seq.length ${serialized} == 4 /\ (forall i. bounded (Seq.index ${simd_unit.values} i) 4)"))]
+#[hax_lib::ensures(|out| {
+    let serialized_future = future(serialized);
+    fstar!(r"
+Seq.length ${serialized_future} == Seq.length ${serialized} /\
+(let inp = bit_vec_of_int_t_array #I32 #(mk_usize 8) ${simd_unit.values} 4 in
+ let out = bit_vec_of_int_t_array #U8  #(mk_usize 4) ${serialized_future} 8 in
+ forall (i: nat {i < 8 * 4}). inp i == out i)
+")
+})]
+pub fn serialize_4(simd_unit: &Coefficients, serialized: &mut [u8]) {
+    // The commitment has coefficients in [0,15] => each coefficient occupies
+    // 4 bits.
+    cloop! {
+        for (i, coefficients) in simd_unit.values.chunks_exact(2).enumerate() {
+            hax_lib::loop_invariant!(|i: usize| {fstar!(r"
+            Seq.length $serialized == 4 /\ (
+              let inp = bit_vec_of_int_t_array #I32 #(mk_usize 8) ${simd_unit.values} 4 in
+              let out = bit_vec_of_int_t_array #U8 #(mk_usize 4) $serialized 8 in
+              forall (n: nat {n < v i * 8}). out n == inp n
+            )")});
+            let r0 = encode_4(coefficients);
```

Comment:
Do we need `r0` as a temporary here? If not, assigning it directly would look nicer.

Reply:
Yes, indeed, that was useful at some point for writing the proof, but it's no longer useful, will drop it!